module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1987 @@
+ * objset_phys_t. The first one is "portable" and is the logical root of the
+ * MAC tree maintianed in the metadnode's bps. The second, is "local" and is
+ * used as the root MAC for the user accounting objects, which are also not
+ * transferred via "zfs send". The portable MAC is sent in the DRR_BEGIN payload
+ * of the send file. The useraccounting code ensures that the useraccounting
+ * info is not present upon a receive, so the local MAC can simply be cleared
+ * out at that time. For more info about objset_phys_t authentication, see
+ * zio_crypt_do_objset_hmacs().
+ *
+ * CONSIDERATIONS FOR DEDUP:
+ * In order for dedup to work, blocks that we want to dedup with one another
+ * need to use the same IV and encryption key, so that they will have the same
+ * ciphertext. Normally, one should never reuse an IV with the same encryption
+ * key or else AES-GCM and AES-CCM can both actually leak the plaintext of both
+ * blocks. In this case, however, since we are using the same plaindata as
+ * well all that we end up with is a duplicate of the original data we already
+ * had. As a result, an attacker with read access to the raw disk will be able
+ * to tell which blocks are the same but this information is already given away
+ * by dedup anyway. In order to get the same IVs and encryption keys for
+ * equivalent blocks of data we use an HMAC of the plaindata. We use an HMAC
+ * here a reproducible checksum of the plaindata is never available to the
+ * attacker. The HMAC key is kept alongside the master key, encrypted on disk.
+ * The first 64 bits of the HMAC are used in place of the random salt, and the
+ * next 96 bits are used as the IV.
+ */
+
+/*
+ * After encrypting many blocks with the same key we may start to run up
+ * against the theoretical limits of how much data can securely be encrypted
+ * with a single key using the supported encryption modes. The most obvious
+ * limitation is that our risk of generating 2 equivalent 96 bit IVs increases
+ * the more IVs we generate (which both GCM and CCM modes strictly forbid).
+ * This risk actually grows surprisingly quickly over time according to the
+ * Birthday Problem. With a total IV space of 2^(96 bits), and assuming we have
+ * generated n IVs with a cryptographically secure RNG, the approximate
+ * probability p(n) of a collision is given as:
+ *
+ * p(n) ~= e^(-n(n-1)/(2*(2^96)))
+ *
+ * [http://www.math.cornell.edu/~mec/2008-2009/TianyiZheng/Birthday.html]
+ *
+ * Assuming that we want to ensure that p(n) never goes over 1 / 1 trillion
+ * we must not write more than 398065730 blocks with the same encryption key,
+ * which is significantly less than the zettabyte of data that ZFS claims to
+ * be able to store. To counteract this, we rotate our keys after 400000000
+ * blocks have been written by generating a new random 64 bit salt for our
+ * HKDF encryption key generation function.
+ */
+unsigned long zfs_key_max_salt_uses = 400000000;
```

Comment:
> I'm assuming we want this to be a tunable rather than a macro so that we can lower it for testing.

It is a tunable :)

> What would you think about making this an int, as a way to ensure that it can't be tuned to a too-dangerous value.

My only issue with that is that (as I understand it) technically the max value for UINT_MAX can be as low as 65535. Maybe a better solution is to cap it at its usage point in the code at 800,000,000 so that the root user can't mess with data security. Thoughts?